pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,1438 @@
+		}
+	}
+}
+
+func (ctrl *PersistentVolumeController) addClaim(obj interface{}) {
+	if !ctrl.isFullySynced() {
+		return
+	}
+
+	claim, ok := obj.(*api.PersistentVolumeClaim)
+	if !ok {
+		glog.Errorf("Expected PersistentVolumeClaim but handler received %+v", obj)
+		return
+	}
+	if err := ctrl.syncClaim(claim); err != nil {
+		glog.Errorf("PersistentVolumeController could not add claim %q: %+v", claimToClaimKey(claim), err)
+	}
+}
+
+func (ctrl *PersistentVolumeController) updateClaim(oldObj, newObj interface{}) {
+	if !ctrl.isFullySynced() {
+		return
+	}
+
+	newClaim, ok := newObj.(*api.PersistentVolumeClaim)
+	if !ok {
+		glog.Errorf("Expected PersistentVolumeClaim but handler received %+v", newObj)
+		return
+	}
+	if err := ctrl.syncClaim(newClaim); err != nil {
+		glog.Errorf("PersistentVolumeController could not update claim %q: %+v", claimToClaimKey(newClaim), err)
+	}
+}
+
+func (ctrl *PersistentVolumeController) deleteClaim(obj interface{}) {
+	if !ctrl.isFullySynced() {
+		return
+	}
+
+	var volume *api.PersistentVolume
+	if pvc, ok := obj.(*api.PersistentVolumeClaim); ok {
+		if pvObj, exists, _ := ctrl.volumes.store.GetByKey(pvc.Spec.VolumeName); exists {
+			if pv, ok := pvObj.(*api.PersistentVolume); ok {
+				volume = pv
+			}
+		}
+	}
+	if unk, ok := obj.(cache.DeletedFinalStateUnknown); ok && unk.Obj != nil {
+		if pv, ok := unk.Obj.(*api.PersistentVolume); ok {
+			volume = pv
+		}
+	}
+
+	// sync the volume when its claim is deleted.  Explicitly sync'ing the
+	// volume here in response to claim deletion prevents the volume from
+	// waiting until the next sync period for its Release.
+	if volume != nil {
+		err := ctrl.syncVolume(volume)
+		if err != nil {
+			glog.Errorf("PersistentVolumeController could not update volume %q from deleteClaim handler: %+v", volume.Name, err)
+		}
+	}
+}
+
+func (ctrl *PersistentVolumeController) syncClaim(claim *api.PersistentVolumeClaim) error {
+	glog.V(5).Infof("Synchronizing PersistentVolumeClaim[%s]: %s", claimToClaimKey(claim), ctrl.getClaimStatusForLogging(claim))
+
+	if !ctrl.hasAnnotation(claim.ObjectMeta, annWasEverBound) {
+		return ctrl.syncUnboundClaim(claim)
+	} else {
+		return ctrl.syncBoundClaim(claim)
+	}
+}
+
+func (ctrl *PersistentVolumeController) syncUnboundClaim(claim *api.PersistentVolumeClaim) error {
+	// This is a new PVC that has not completed binding
+	// OBSERVATION: pvc is "Pending"
+	if claim.Spec.VolumeName == "" {
+		// User did not care which PV they get.
+		// [Unit test set 1]
+		volume, err := ctrl.volumes.findBestMatchForClaim(claim)
+		if err != nil {
+			glog.V(2).Infof("Synchronizing unbound PersistentVolumeClaim[%s]: Error finding PV for claim: %v", claimToClaimKey(claim), err)
+			return fmt.Errorf("Error finding PV for claim %q: %v", claimToClaimKey(claim), err)
+		}
+		if volume == nil {
+			glog.V(5).Infof("Synchronizing unbound PersistentVolumeClaim[%s]: no volume found", claimToClaimKey(claim))
+			// No PV could be found
```

Comment:
Will `findBestMatchForClaim` ignore claims with this annotation set?  Or, are they opportunistically matched against existing volumes that might satisfy them, and then new volumes provisioned if there are none?  This is the place to say it.